File: irrgarten/Labyrinth.java
package irrgarten;

import java.util.ArrayList;

public class Labyrinth {
    private static final char BLOCK_CHAR = 'X';
    private static final char MONSTER_CHAR = 'M';
    private static final char EMPTY_CHAR = '-';
    private static final char COMBAT_CHAR = 'C';
    private static final char EXIT_CHAR = 'E';
    private static final int ROW = 0;
    private static final int COL = 1;
    private final int nRows;
    private final int nCols;
    private int exitRow;
    private int exitCol;

    private Monster[][] monsters;
    private char[][] labyrinth;
    private Player[][] players;

    public Labyrinth(int nRows, int nCols, int exitRow, int exitCol){
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitRow = exitRow;
        this.exitCol = exitCol;
        monsters = new Monster[nRows][nCols];
        players = new Player[nRows][nCols];
        labyrinth = new char[nRows][nCols];

        for(int i=0; i < nRows; i++){
            for(int j=0; j < nCols; j++){
                labyrinth[i][j] = EMPTY_CHAR;
            }
        }

        labyrinth[exitRow][exitCol] = EXIT_CHAR;
    }

    public int getnRows() {
        return this.nRows;
    }

    public int getnCols() {
        return this.nCols;
    }

    public void setExit(int row, int col) {
        this.exitRow = row;
        this.exitCol = col;
    }

    public void spreadPlayers(ArrayList<Player> players){
        for (Player p : players) {
            int[] pos = this.randomEmptyPos();
            this.putPlayer2D(-1, -1, pos[0], pos[1], p);
        }
    }

    public boolean haveAWinner(){
        return players[exitRow][exitCol] != null;
    }


    public String toString(){

        String completeMatrixString = "";

        for (int i = 0; i < nRows; i++){
            completeMatrixString += "| ";
            for (int j = 0; j < nCols; j++){
                completeMatrixString += labyrinth[i][j] + " | ";
            }
            completeMatrixString += "\n";
        }

        return completeMatrixString;
    }

    public void addMonster(int row, int col, Monster monster){
        if(this.posOK(row, col) && this.emptyPos(row, col)){
            labyrinth[row][col] = MONSTER_CHAR;
            monsters[row][col] = monster;
            monster.setPos(row, col);
        }
    }

    public Monster putPlayer(Directions direction, Player player){
        int oldRow = player.getRow();
        int oldCol = player.getCol();

        int[] newPos = dir2Pos(oldRow, oldCol, direction);

        return putPlayer2D(oldRow, oldCol, newPos[0], newPos[1], player);
    }

    public void addBlock(Orientation orientation, int startRow, int startCol, int length){
        int incRow = 0;
        int incCol = 1;

        if (orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        }

        int row = startRow;
        int col = startCol;

        while (this.posOK(row, col) && (this.emptyPos(row, col) && (length > 0))){
            labyrinth[row][col] = BLOCK_CHAR;
            length--;
            row += incRow;
            col += incCol;
        }
    }

    public ArrayList<Directions> validMoves(int row, int col){
        ArrayList<Directions> output = new ArrayList<>();

        if (this.canStepOn(row+1, col)){
            output.add(Directions.DOWN);
        }
        if (this.canStepOn(row-1, col)){
            output.add(Directions.UP);
        }
        if (this.canStepOn(row, col+1)){
            output.add(Directions.RIGHT);
        }
        if (this.canStepOn(row, col-1)){
            output.add(Directions.LEFT);
        }

        return output;
    }

    private boolean posOK(int row, int col){
        return (row < nRows) && ( col < nCols);
    }

    private boolean emptyPos(int row, int col){
        return labyrinth[row][col] == EMPTY_CHAR;
    }

    private boolean monsterPos(int row, int col){
        return labyrinth[row][col] == MONSTER_CHAR;
    }

    private boolean exitPos(int row, int col){
        return labyrinth[row][col] == EXIT_CHAR;
    }

    private boolean combatPos(int row, int col){
        return labyrinth[row][col] == COMBAT_CHAR;
    }

    private boolean canStepOn(int row, int col){
        //No pongo && posOK porque ninguna de esas opciones dará true en una posición invalida)
        return (this.monsterPos(row, col) || this.emptyPos(row, col) || this.exitPos(row, col)) && this.posOK(row,col);
    }

    private void updateOldPos(int row, int col){
        if(this.posOK(row, col)){
            if(this.combatPos(row, col)){
                labyrinth[row][col] = MONSTER_CHAR;
            }
            else{
                labyrinth[row][col] = EMPTY_CHAR;
            }
        }
    }

    private int[] dir2Pos(int row, int col, Directions direction){
        int[] nextPosition = new int[2];

        switch (direction) {
            case UP:
                nextPosition[0] = row-1;
                nextPosition[1] = col;
                break;
            case DOWN:
                nextPosition[0] = row+1;
                nextPosition[1] = col;
                break;
            case LEFT:
                nextPosition[0] = row;
                nextPosition[1] = col-1;
                break;
            case RIGHT:
                nextPosition[0] = row;
                nextPosition[1] = col+1;
                break;
        }

        return nextPosition;
    }

    private int[] randomEmptyPos(){
        int[] randomPos = new int[2];

        do {
            randomPos[0] = Dice.randomPos(nRows-1);
            randomPos[1] = Dice.randomPos(nCols-1);
        } while (!this.posOK(randomPos[0], randomPos[1]));

        return randomPos;
    }

    private Monster putPlayer2D(int oldRow, int oldCol, int row, int col, Player player){
        Monster output = null;

        if (canStepOn(row, col)) {
            if (posOK(oldRow, oldCol)) {
                Player p = players[oldRow][oldCol];

                if (p == player) {
                    this.updateOldPos(oldRow, oldCol);
                    players[oldRow][oldCol] = null;
                }
            }

            if (this.monsterPos(row, col)) {
                labyrinth[row][col] = COMBAT_CHAR;
                output = monsters[row][col];
            }
            else{
                labyrinth[row][col] = (char) player.getNumber();
            }

            players[row][col] = player;
            player.setPos(row, col);
        }

        return output;
    }

}

File: irrgarten/Labyrinth.rb
module Irrgarten
  require_relative 'Dice'
  require_relative 'Monster'
  class Labyrinth
    @@BLOCK_CHAR='X'
    @@EMPTY_CHAR='-'
    @@MONSTER_CHAR='M'
    @@COMBAT_CHAR='C'
    @@EXIT_CHAR='E'
    @@ROW = 0
    @@COL= 1


    #Constructor
    def initialize(nRows,nCols,exitRow,exitCol)
      @nRows = nRows
      @nCols= nCols
      @exitRow = exitRow
      @exitCol = exitCol
      @monsters = Array.new(@nRows) { Array.new(@nCols){nil} }
      @labyrinth = Array.new(@nRows) { Array.new(@nCols){@@EMPTY_CHAR} }
      @labyrinth[exitRow][exitCol] = @@EXIT_CHAR
      @players = Array.new(@nRows) { Array.new(@nCols){nil} }
    end
    #Por implementar
    def spreadPlayers(players)
      for i in 0..(players.size-1)
        p=players[i]
        pos=self.randomEmptyPos
        self.putPlayer2D(-1,-1,pos[@@ROW],pos[@@COL],p)
      end
    end
    #Dice si hay un ganador
    def haveAWinner
      return @players[@exitRow][@exitCol]!=nil
    end
    #Imprime el laberinto
    def to_s
      resultado = ""

      @labyrinth.each_with_index do |r, r_index|
        r.each_with_index do |c, c_index|
          resultado += (@labyrinth[r_index][c_index]).to_s
          resultado +=" "
        end
        resultado+="\n"
      end
      return resultado
    end
    #Añade un monstruo
    def addMonster(row,col,monster)
      if posOK(row,col) && @labyrinth[row][col]==@@EMPTY_CHAR && monster!= nil   #no muy seguro de la última condición
        @labyrinth[row][col]=@@MONSTER_CHAR
        @monsters[row][col]=monster
        monster.setPos(row,col)
      end
    end
    #Por implementar
    def putPlayer(direction,player)
      oldRow=player.getRow
      oldCol=player.getCol
      newPos=dir2Pos(oldRow,oldCol,direction)
      monster=putPlayer2D(oldRow,oldCol,newPos[0],newPos[1],player)
      monster
    end
    #Por implementar
    def addBlock(orientation,startRow,startCol,length)
      incRow=0
      incCol=0
      if orientation==Orientation::VERTICAL
        incRow=1
        incCol=0
      else
        incRow=0
        incCol=1
      end
      while self.posOK(startRow,startCol) &&self.emptyPos(startRow,startCol)&&length>0
        @labyrinth[startRow][startCol]=@@BLOCK_CHAR
        length-=1
        startRow+=incRow
        startCol+=incCol
      end
    end
    #POr implementar
    def validMoves(row,col)
      output=[]
      if self.canStepOn(row+1,col)
        output<<Directions::DOWN
      end
      if self.canStepOn(row-1,col)
        output<<Directions::UP
      end
      if self.canStepOn(row,col+1)
        output<<Directions::RIGHT
      end
      if self.canStepOn(row,col-1)
        output<<Directions::LEFT
      end
      return output
    end
    private
    #Dice se la posición es correcta
    def posOK(row,col)
      if row < @nRows && row >=0 && col <@nCols && col>=0
        return true
      else
        return false
      end
    end
    #Dice si la posición está vacia
    def emptyPos(row,col)
      return @labyrinth[row][col]==@@EMPTY_CHAR
    end
    #Dice si hay un monstruo en la posicion
    def monsterPos(row,col)
      return @labyrinth[row][col]==@@MONSTER_CHAR
    end
    #DIce si es la casilla de salida.
    def exitPos(row,col)
      return @labyrinth[row][col]==@@EXIT_CHAR
    end
    #Dice si es una posicion de combate
    def combatPos(row,col)
      return (@labyrinth[row][col]==@@MONSTER_CHAR and @labyrinth[row][col]==@@COMBAT_CHAR)
    end
    #Dice si puedes avanzar a esa posicion
    def canStepOn(row,col)
      return (posOK(row,col) && (emptyPos(row,col) || monsterPos(row,col) || exitPos(row,col)))
    end
    #Actualiza la posición
    def updateOldPos(row,col)
      if @labyrinth[row][col]==@@COMBAT_CHAR
        @labyrinth[row][col]=@@MONSTER_CHAR
      else
        @labyrinth[row][col]=@@EMPTY_CHAR
      end
    end
    #Dice a la posición a la que se avanzaría si se mueve
    def dir2Pos(row, col ,direction)
      pos = [0, 0]

      case direction
      when Directions::LEFT
        pos[@@ROW] = row
        pos[@@COL] = col -1
      when Directions::RIGHT
        pos[@@ROW] = row
        pos[@@COL] = col +1
      when Directions::UP
        pos[@@ROW] = row -1
        pos[@@COL] = col
      when Directions::DOWN
        pos[@@ROW] = row +1
        pos[@@COL] = col 
      end
      return pos
    end
    #Da una posicion aleatoria
    def randomEmptyPos()
      pos = [0, 0]
      found = false
      while !found
        pos[@@ROW] = Dice.randomPos(@nRows-1)
        pos[@@COL] = Dice.randomPos(@nCols-1)
        found = @labyrinth[pos[@@ROW]][pos[@@COL]] == @@EMPTY_CHAR
      end
      return pos
    end
    #Por implementar
    def putPlayer2D(oldRow,oldCol,row,col,player)
      output=nil
      if canStepOn(row,col)
        if posOK(oldRow,oldCol)
          p=@players[oldRow][oldCol]
          if p==player
            updateOldPos(oldRow,oldCol)
            @players[oldRow][oldCol]=nil
          end
        end
        monsterPos=monsterPos(row,col)
        if monsterPos
          @labyrinth[row][col]=@@COMBAT_CHAR
          output=@monsters[row][col]
        else
          number=player.getNumber
          #probablemte no salga bien el número en el tablero
          @labyrinth[row][col]=number
        end
        @players[row][col]=player
        player.setPos(row,col)
      end
      return output
    end
  end
end